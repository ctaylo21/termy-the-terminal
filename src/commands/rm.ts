import { CommandResponse, FileSystem } from '../index';
import cloneDeep from 'lodash/cloneDeep';
import get from 'lodash/get';
import has from 'lodash/has';
import unset from 'lodash/unset';
import { getInternalPath } from './utilities/index';

/**
 * Deletes path from given filesystem and returns updatd filesystem
 * without modifying original arg.
 *
 * @param fileSystem {object} - filesystem to act upon
 * @param pathToDelete {string} - internally-formatted path to delete
 */
function handleDelete(
  fileSystem: FileSystem,
  pathToDelete: string,
): CommandResponse {
  const newFileSystem = cloneDeep(fileSystem);
  unset(newFileSystem, pathToDelete);
  return {
    updatedState: {
      fileSystem: newFileSystem,
    },
  };
}

/**
 * Given a path, removes the object at that location if possible. Rejects if
 * the parameters aren't correct for the given item
 *
 * @param fileSystem {object} - filesystem to act upon
 * @param currentPath {string} - current path within filesystem
 * @param folderPath  {string} - path of object to remove
 * @returns Promise<object> - resolves if rm was successful, rejects if not
 */
export default function rm(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string,
  options?: string,
): Promise<CommandResponse> {
  return new Promise((resolve, reject): void => {
    if (!targetPath) {
      reject('Missing argument to rm');
    }

    const internalCdPath = getInternalPath(currentPath, targetPath);

    if (has(fileSystem, internalCdPath)) {
      if (get(fileSystem, internalCdPath).type === 'FOLDER') {
        if (options === '-r') {
          resolve(handleDelete(fileSystem, internalCdPath));
        } else {
          reject(`Can't remove ${targetPath}. It is a directory.`);
        }
      }

      if (get(fileSystem, internalCdPath).type === 'FILE') {
        resolve(handleDelete(fileSystem, internalCdPath));
      }
    }

    reject(`Can't remove ${targetPath}. No such file or directory.`);
  });
}
