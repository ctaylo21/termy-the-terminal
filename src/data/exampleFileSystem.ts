const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          video1: {
            type: 'FILE',
          },
        },
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
};

export default exampleFileSystem;
