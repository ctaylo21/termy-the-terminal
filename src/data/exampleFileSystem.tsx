import React from 'react';
import { FileSystem } from '../index';
import dogImg from '../../src/images/dog.png';

interface BlogPostProps {
  date: string;
  content: string;
}

export const BlogPost: React.FC<BlogPostProps> = ({
  date,
  content,
}): JSX.Element => (
  <>
    <h3>{date}</h3>
    <p>{content}</p>
  </>
);

const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          file2: {
            type: 'FILE',
            content: 'Contents of file 2',
            extension: 'txt',
          },
        },
      },
      dog: {
        type: 'FILE',
        content: dogImg,
        extension: 'png',
      },
      file1: {
        type: 'FILE',
        content: 'Contents of file 1',
        extension: 'txt',
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
  file3: {
    type: 'FILE',
    content: 'Contents of file 3',
    extension: 'txt',
  },
  blog: {
    type: 'FILE',
    content: <BlogPost date="3/22" content="Today is a good day" />,
    extension: 'txt',
  },
};

export default exampleFileSystem;
