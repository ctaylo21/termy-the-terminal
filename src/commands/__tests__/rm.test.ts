import rm from '../rm';
const { handler, autoCompleteHandler } = rm;
import exampleFileSystem from '../../data/exampleFileSystem';
import { FileSystem } from '../../index';
import cloneDeep from 'lodash/cloneDeep';

describe('rm suite', (): void => {
  describe('success', (): void => {
    describe('non-folders', (): void => {
      it('should remove a file with no options', async (): Promise<void> => {
        const expectedFileSystem = cloneDeep(exampleFileSystem);
        delete expectedFileSystem.file3;

        return expect(
          handler(exampleFileSystem, '/', 'file3.txt'),
        ).resolves.toEqual({
          updatedState: {
            fileSystem: expectedFileSystem,
          },
        });
      });

      it('should remove a nested file', async (): Promise<void> => {
        const expectedFileSystem = cloneDeep(exampleFileSystem);
        delete ((expectedFileSystem.home.children as FileSystem).videos
          .children as FileSystem).file2;

        return expect(
          handler(exampleFileSystem, '/', 'home/videos/file2.txt'),
        ).resolves.toEqual({
          updatedState: {
            fileSystem: expectedFileSystem,
          },
        });
      });

      it('should remove file in parent path', async (): Promise<void> => {
        const expectedFileSystem = cloneDeep(exampleFileSystem);
        delete expectedFileSystem.file3;

        return expect(
          handler(exampleFileSystem, '/home/videos', '../../file3.txt'),
        ).resolves.toEqual({
          updatedState: {
            fileSystem: expectedFileSystem,
          },
        });
      });
    });

    describe('folders', (): void => {
      it('should delete at same level', async (): Promise<void> => {
        const expectedFileSystem = cloneDeep(exampleFileSystem);
        delete expectedFileSystem.docs;

        return expect(
          handler(exampleFileSystem, '/', 'docs', '-r'),
        ).resolves.toEqual({
          updatedState: {
            fileSystem: expectedFileSystem,
          },
        });
      });

      it('should delete from nested path', async (): Promise<void> => {
        const expectedFileSystem = cloneDeep(exampleFileSystem);
        delete (expectedFileSystem.home.children as FileSystem).videos;

        return expect(
          handler(exampleFileSystem, '/', 'home/videos', '-r'),
        ).resolves.toEqual({
          updatedState: {
            fileSystem: expectedFileSystem,
          },
        });
      });

      it('should delete from parent path', async (): Promise<void> => {
        const expectedFileSystem = cloneDeep(exampleFileSystem);
        delete expectedFileSystem.docs;

        return expect(
          handler(exampleFileSystem, '/home/videos', '../../docs', '-r'),
        ).resolves.toEqual({
          updatedState: {
            fileSystem: expectedFileSystem,
          },
        });
      });
    });
  });

  describe('failure', (): void => {
    it('should reject if path is invalid', async (): Promise<void> => {
      return expect(
        handler(exampleFileSystem, '/', 'invalid'),
      ).rejects.toMatchSnapshot();
    });

    it('should reject if no target path provided', async (): Promise<void> => {
      return expect(
        handler(exampleFileSystem, '/', ''),
      ).rejects.toMatchSnapshot();
    });

    it('should reject if no options and target is a folder', async (): Promise<
      void
    > => {
      return expect(
        handler(exampleFileSystem, '/', 'home'),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('auto complete', (): void => {
    test('empty value', async (): Promise<void> => {
      const { commandResult } = await autoCompleteHandler(
        exampleFileSystem,
        '/home/user',
        '',
      );

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(Object.keys(commandResult!)).toContain('test');
    });

    test('should filter single level target', async (): Promise<void> => {
      const { commandResult } = await autoCompleteHandler(
        exampleFileSystem,
        '/',
        'fi',
      );

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('file3.txt');
      expect(autoCompleteValues).toContain('file4.txt');
      expect(autoCompleteValues).not.toContain('blog');
      expect(autoCompleteValues).not.toContain('docs');
      expect(autoCompleteValues).not.toContain('home');
    });

    test('invalid path should return nothing', async (): Promise<void> => {
      const lsResult = await autoCompleteHandler(
        exampleFileSystem,
        '/bad/path',
        '',
      );

      expect(lsResult.commandResult).toBeUndefined();
    });

    test('relative path', async (): Promise<void> => {
      const { commandResult } = await autoCompleteHandler(
        exampleFileSystem,
        '/',
        'home/fi',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('file1.txt');
      expect(autoCompleteValues).toContain('file5.txt');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('dog.png');
    });

    test('relative path with dotdot', async (): Promise<void> => {
      const { commandResult } = await autoCompleteHandler(
        exampleFileSystem,
        '/',
        'home/../home/fi',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('file1.txt');
      expect(autoCompleteValues).toContain('file5.txt');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('dog.png');
    });

    test('absolute path', async (): Promise<void> => {
      const { commandResult } = await autoCompleteHandler(
        exampleFileSystem,
        '/',
        '/home/d',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('dog.png');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('file1.txt');
      expect(autoCompleteValues).not.toContain('file5.txt');
    });

    test('absolute path with ..', async (): Promise<void> => {
      const { commandResult } = await autoCompleteHandler(
        exampleFileSystem,
        '/',
        '/home/user/../d',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('dog.png');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('file1.txt');
      expect(autoCompleteValues).not.toContain('file5.txt');
    });
  });
});
