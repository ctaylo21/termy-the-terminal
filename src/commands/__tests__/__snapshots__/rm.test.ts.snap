// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rm suite failure should reject if no options and target is a folder 1`] = `"Can't remove home. It is a directory."`;

exports[`rm suite failure should reject if no target path provided 1`] = `"Missing argument to rm"`;

exports[`rm suite failure should reject if path is invalid 1`] = `"Can't remove invalid. No such file or directory."`;
